hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/update/server/SCMUpdateClientManager.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with this
+ * work for additional information regarding copyright ownership.  The ASF
+ * licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package org.apache.hadoop.hdds.scm.update.server;
+
+import org.apache.hadoop.hdds.scm.exceptions.SCMException;
+import org.apache.ratis.thirdparty.io.grpc.Status;
+import org.apache.ratis.thirdparty.io.grpc.stub.StreamObserver;
+import org.apache.hadoop.hdds.protocol.scm.proto.SCMUpdateServiceProtos.Type;
+import org.apache.hadoop.hdds.protocol.scm.proto.SCMUpdateServiceProtos.UpdateRequest;
+import org.apache.hadoop.hdds.protocol.scm.proto.SCMUpdateServiceProtos.UpdateResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Class that manages SCM update clients.
+ */
+public class SCMUpdateClientManager {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(SCMUpdateClientManager.class);
+  private Map<UUID, SCMUpdateClientInfo> clients;
+  private Map<Type, SCMUpdateHandler> handlers;
+
+  public SCMUpdateClientManager() {
+    clients = new ConcurrentHashMap<>();
+    handlers = new ConcurrentHashMap<>();
+  }
+
+  public void registerHandler(SCMUpdateHandler handler) {
+    handlers.put(handler.getType(), handler);
+  }
+
+  public void unRegisterHandler(Type type) {
+    handlers.remove(type);
+  }
+
+  public UUID addClient() throws SCMException {
+    UUID clientId = UUID.randomUUID();
+    int retryCount = 5;
+    while (clients.containsKey(clientId)) {
+      if (retryCount > 0) {
+        clientId = UUID.randomUUID();
+        retryCount--;
+      } else {
+        throw new SCMException("Failed to add CRL client with random clientId" +
+            " collision", SCMException.ResultCodes.FAILED_TO_ADD_CRL_CLIENT);
+      }
+    }
+
+    SCMUpdateClientInfo clientInfo = new SCMUpdateClientInfo(clientId);
+    clients.put(clientId, clientInfo);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **59 and 70**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).